File: mars-rover-functional/rover.ts
type PowerMode = "LOW_POWER" | "NORMAL"
type Position = number
type Watts = number
type MessageName = string

type Rover = {
    position: Position,
    mode: PowerMode
    generatorWatts: Watts
}

interface Result {completed: boolean, roverStatus?: Rover} 
type Command = ["STATUS_CHECK"] | ["MODE_CHANGE", PowerMode] | ["MOVE", Position]

type Message = {
    name: MessageName,
    commands: Array<Command>
}

type MessageResponse = {
    message: MessageName,
    results: Array<Result>
}

const receiveMessage = (message: Message, rover: Rover): [MessageResponse, Rover] => {
    const [results, newRover] = processCommands(message.commands, [], rover)
    return [{message: message.name, results: results}, newRover]
}

const processCommands = (cmds: Array<Command>, results: Array<Result>, rover: Rover): [Array<Result>, Rover] => {
    if (cmds.length === 0) {
        return [results, rover]
    }
    const [cmd, ...remainingCommands] = cmds
    const [result, newRover] = processCommand(cmd, rover)
    return processCommands(remainingCommands, [...results, result], newRover)
}


const processCommand = (cmd: Command, rover: Rover): [Result, Rover] => {
    const [cmdType, value] =  cmd
    if (cmdType === "MODE_CHANGE") {
        return [{completed: true}, {...rover, mode: value} ]
    }

    if (cmdType === "MOVE" && rover.mode === "LOW_POWER") {
        return [{completed: false}, rover]
    }

    if (cmdType === "MOVE" && rover.mode === "NORMAL") {
        return [{completed: true}, {...rover, position: value}]
    }

    return [{completed: true, roverStatus: rover}, rover]
}

File: mars-rover-functional/rover.hs
#! /usr/bin/env stack
--stack --resolver lts-16.26 script
data PowerMode = LowPower | Normal deriving (Show, Eq)
type Position = Integer
type Watts = Integer
type MessageName = String
type RoverStatus = String

data Rover = Rover { position :: Position
                   , mode :: PowerMode
                   , generatorWatts :: Watts
                   } deriving Show
 
data Completed = Failed | Succeeded deriving Show
data Result = Result Completed (Maybe RoverStatus) deriving Show
data Command = ModeChange PowerMode | Move Position | StatusCheck deriving Show
data Message = Message MessageName [Command] deriving Show
data Response = Response MessageName [Result] deriving Show

receiveMessage :: Message -> Rover -> (Response, Rover)
receiveMessage (Message messageName commands) rover = 
    (Response messageName results, newRover)
    where (results, newRover) = processCommands commands [] rover


processCommands :: [Command] -> [Result] -> Rover -> ([Result], Rover)
processCommands [] results rover = (results, rover)
processCommands (cmd: cmds) results rover = processCommands cmds (results ++ [result]) newRover
    where (result, newRover) = processCommand cmd rover


processCommand :: Command -> Rover -> (Result, Rover)
processCommand (ModeChange newMode) (Rover ogPosition _ ogWatts ) = (Result Succeeded Nothing, Rover ogPosition newMode ogWatts)
processCommand StatusCheck rover = (Result Succeeded (Just $ show rover), rover)
processCommand (Move newPosition) rover@(Rover ogPosition LowPower watts) = (Result Failed Nothing, rover)
processCommand (Move newPosition) (Rover _ Normal watts) = (Result Succeeded Nothing, Rover newPosition Normal watts)



main :: IO ()
main = do 
    let commands = [ StatusCheck
                   , ModeChange LowPower
                   , StatusCheck
                   , Move 300
                   , StatusCheck
                   , ModeChange Normal
                   , StatusCheck
                   , Move 300
                   , StatusCheck
                   ]
    let message = Message "Test" commands
    let rover = Rover 100 Normal 110
    let (response, updatedRover) = receiveMessage message rover
    print response
    print updatedRover
    